src/range_query/sqrt_decomp.rs

```diff
@@ -129,14 +131,54 @@ impl PiecewiseLinearFn {
     fn update_envelope(&mut self) {
         self.recent_lines.extend(self.sorted_lines.drain(..));
         self.recent_lines.sort_unstable();
-        for (slope, intercept) in self.recent_lines.drain(..) {
-            // TODO: do convex hull trick algorithm
-            self.sorted_lines.push((slope, intercept));
+        self.intersections.clear();
+
+        'outer: for (m1, b1) in self.recent_lines.drain(..) {
+            while let Some(&(m2, b2)) = self.sorted_lines.last() {
+                // The lines are sorted by intercept so an incoming line with
+                // the same slope can never beat the previous line.
+                if m2 == m1 {
+                    continue 'outer;
+                }
+                let new_intersection = (b1 - b2) as f64 / (m2 - m1) as f64;
+                if let Some(&last_intersection) = self.intersections.last() {
+                    if new_intersection < last_intersection {
+                        self.intersections.pop();
+                        self.sorted_lines.pop();
+                        continue;
+                    }
+                }
+                self.intersections.push(new_intersection);
+                break;
+            }
+            self.sorted_lines.push((m1, b1));
+        }
+    }
```

Comment:
I don't like all the breaks and continues, but there's a couple annoying edge cases that make it hard to do it in an elegant way. Maybe I'll refactor

Reply:
Something like "let Some x = y and pred(x)" would be nice but I don't think it's in the language. Probably a straightup "match" would be better

Reply:
You mean chaining like this? https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md
I guess it will be added to the language eventually; I can combine the nested `if`s at that time.